mrmustard/utils/settings.py

```diff
@@ -122,6 +124,14 @@ def __init__(self):
         self.ATOL: float = 1e-8
         "The absolute tolerance when comparing two values or arrays. Default is 1e-8."
 
+        self._frozen = True
+
+    def __setattr__(self, name, value):
+        """Once the class is initialized, do not allow the addition of new settings."""
+        if self._frozen and name != "HBAR" and not hasattr(self, name):
```

Comment:
Gah maybe instead of locking HBAR we just warn users that when they change it unexpected behaviour can occur?

Reply:
my goal was to not mess with how things were in this PR, but I'm open to discussing new plans for HBAR. I'm guessing the lock was added because people were accidentally doing sketchy math

Reply:
I think one reason was that some of functions for plotting states use cached arrays that depend on the value of hbar

Reply:
Yeah, the way the HBAR lock is setup right now kind of conflicts with the settings lock hence requiring this attribute specific logic (which I'd like to avoid as much as possible). I don't see an easy way to fix this which is why I suggested a warning instead. I'd like to say that just telling a user "warning: changing this setting may conflict with any prior computations" is sufficient but I'm not too sure

Reply:
I'm fine with a warning or an error, but this explicit HBAR check is required either way to ensure the warning/error doesn't appear on access, right?

Reply:
Not necessarily? Like if we do 
```suggestion
    @property
    def HBAR(self):
        r"""The value of the Planck constant. Default is ``2``.
        Cannot be changed after its value is queried for the first time.
        """
        return self._hbar
```
```suggestion
    @HBAR.setter
    def HBAR(self, value: float):
        warnings.warn("Changing HBAR can conflict with prior computations.") 
        self._hbar = value
```
then we only get warnings when someone changes HBAR. (Tested locally)

Reply:
oh I see. I was assuming we want to avoid the warning if it's their first time changing it, but I kinda agree that this warning is fine if @elib20 is ok with it

Reply:
ok i've updated it to just raise a warning whenever set